File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/searching/binary_search_tree.py
class Node:
    def __init__(self, value: int) -> None:
        self.value = value
        self.left: Node | None = None
        self.right: Node | None = None


class BinarySearchTree:
    def __init__(self) -> None:
        self.root: Node | None = None

    def insert(self, value: int) -> None:
        self.root = self._insert(self.root, value)

    def contains(self, value: int) -> bool:
        current = self.root
        while current is not None:
            if value == current.value:
                return True
            current = current.left if value < current.value else current.right
        return False

    def inorder(self) -> list[int]:
        result: list[int] = []
        self._inorder(self.root, result)
        return result

    def _insert(self, node: Node | None, value: int) -> Node:
        if node is None:
            return Node(value)
        if value < node.value:
            node.left = self._insert(node.left, value)
        elif value > node.value:
            node.right = self._insert(node.right, value)
        return node

    def _inorder(self, node: Node | None, result: list[int]) -> None:
        if node is None:
            return
        self._inorder(node.left, result)
        result.append(node.value)
        self._inorder(node.right, result)


if __name__ == "__main__":
    sample = [42, 7, 19, 3, 99, 1]
    target = 19
    tree = BinarySearchTree()
    for value in sample:
        tree.insert(value)

    print("Binary Search Tree")
    print(f"input  : {sample}")
    print(f"inorder: {tree.inorder()}")
    print(f"target : {target}")
    print(f"found  : {tree.contains(target)}")

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/searching/BinarySearchTree.java
package com.sortingsearching.searching;

import java.util.ArrayList;
import java.util.List;

public final class BinarySearchTree {
    private Node root;

    public void insert(int value) {
        root = insert(root, value);
    }

    public boolean contains(int value) {
        Node current = root;
        while (current != null) {
            if (value == current.value) {
                return true;
            }
            current = value < current.value ? current.left : current.right;
        }
        return false;
    }

    public List<Integer> inorder() {
        List<Integer> result = new ArrayList<>();
        inorder(root, result);
        return result;
    }

    private Node insert(Node node, int value) {
        if (node == null) {
            return new Node(value);
        }
        if (value < node.value) {
            node.left = insert(node.left, value);
        } else if (value > node.value) {
            node.right = insert(node.right, value);
        }
        return node;
    }

    private void inorder(Node node, List<Integer> result) {
        if (node == null) {
            return;
        }
        inorder(node.left, result);
        result.add(node.value);
        inorder(node.right, result);
    }

    private static final class Node {
        private final int value;
        private Node left;
        private Node right;

        private Node(int value) {
            this.value = value;
        }
    }

    public static void main(String[] args) {
        int[] sample = {42, 7, 19, 3, 99, 1};
        int target = 19;
        BinarySearchTree tree = new BinarySearchTree();
        for (int value : sample) {
            tree.insert(value);
        }

        System.out.println("Binary Search Tree");
        System.out.println("input  : [42, 7, 19, 3, 99, 1]");
        System.out.println("inorder: " + tree.inorder());
        System.out.println("target : " + target);
        System.out.println("found  : " + tree.contains(target));
    }
}
